---
author: Krtin Juneja (KJUNEJA@falcon.bentley.edu)
---

We create a hypothetical repeated measures dataset where the 5 subjects undergo all 4 skin treatments and their rating of the treatment is measured.

In [4]:
import pandas as pd
df = pd.DataFrame( {
    'Subject':        [1,1,1,1,2,2,2,2,3,3,3,3,4,4,4,4,5,5,5,5],
    'Skin Treatment': ['W','X','Y','Z','W','X','Y','Z','W','X',
                       'Y','Z','W','X','Y','Z','W','X','Y','Z'],
    'Rating':         [7,5,8,4,8,10,7,5,7,6,5,4,7,7,4,5,8,8,6,6]
} )
df.head()

Subject Skin Treatment  Rating
0        1              W       7
1        1              X       5
2        1              Y       8
3        1              Z       4
4        2              W       8

Before we conduct a repeated measures ANOVA, we need to decide which approach to use - Univariate or Multivariate. We decide this using Mauchly's test of sphericity. If we fail to reject the null hypothesis then we use the univariate approach. 

 * $H_0 =$ the sphericity assumption holds
 * $H_A =$ the sphericity assumption is violated

We use the `pingouin` statistics package to conduct the test.
Most of the parameters below are self-explanatory, except that `dv` stands for dependent variable.

In [5]:
import pingouin as pg
pg.sphericity( dv='Rating', within='Skin Treatment', subject='Subject', method='mauchly', data=df )

SpherResults(spher=True, W=0.06210054956238558, chi2=7.565056754547507, dof=5, pval=0.20708214225927316)

Since the $p$ value of `skin_treatment` is about $0.2071$, we fail to reject the sphericity assumption at a 5% significance level and use the univariate approach to conduct the repeated measures ANOVA.

In [6]:
# Compute a repeated measures ANOVA using a function pingouin adds to our DataFrame:
df.rm_anova( dv='Rating', within='Skin Treatment', subject='Subject', detailed=False )

Source  ddof1  ddof2         F     p-unc      np2       eps
0  Skin Treatment      3     12  5.117647  0.016501  0.56129  0.541199

Since the $p$ value of about $0.017$ is less that the 5% significance level, we conclude that there is significant evidence of a treatment effect.

Note: If there is more than 1 repeated measures factor, you can add a list of them to the `within` parameter and conduct the test.